File: src/main/java/com/tyiu/corn/controller/ProjectController.java
package com.tyiu.corn.controller;


import com.tyiu.corn.config.exception.NotFoundException;
import com.tyiu.corn.model.dto.ProjectDTO;
import com.tyiu.corn.model.entities.ProjectRequest;
import com.tyiu.corn.model.entities.ProjectInvitation;
import com.tyiu.corn.model.responses.AuthenticationResponse;
import com.tyiu.corn.model.responses.InfoResponse;
import com.tyiu.corn.service.ProjectService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.security.Principal;

@RestController
@RequestMapping("/api/v1/project")
@RequiredArgsConstructor
public class ProjectController {

    private final ProjectService projectService;

    ///////////////////////
    //  _____   ____ ______
    // / ___/  / __//_  __/
    /// (_ /  / _/   / /
    //\___/  /___/  /_/
    ///////////////////////

    @GetMapping("/all")
    public Flux<ProjectDTO> getAllProject(){
        return projectService.getAll()
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    @GetMapping("/get/{projectId}")
    public Mono<ProjectDTO> getProject(@PathVariable Long projectId){
        return projectService.getProject(projectId)
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    @GetMapping("/invites")
    public Flux<ProjectInvitation> getProjectInvitations(Principal principal){
        return projectService.getProjectInvitations(Long.valueOf(principal.getName()))
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    @GetMapping("/request/{projectId}")
    public Flux<ProjectRequest> getProjectRequests(@PathVariable Long projectId){
        return projectService.getProjectRequests(projectId)
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    //////////////////////////////
    //   ___   ____    ____ ______
    //  / _ \ / __ \  / __//_  __/
    // / ___// /_/ / _\ \   / /
    ///_/    \____/ /___/  /_/
    //////////////////////////////

    @PostMapping("/create")
    public Mono<ProjectDTO> createProject(@RequestBody ProjectDTO projectDTO){
        return projectService.createProject(projectDTO)
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    @PostMapping("/send/invite/{projectId}")
    public Mono<ProjectInvitation> sendInvitation(@RequestBody AuthenticationResponse invitation, @PathVariable Long projectId){
        return projectService.sendInvitation(invitation.getId(), projectId)
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    @PostMapping("/send/request/{projectId}")
    public Mono<ProjectRequest> sendApplication(Principal principal, @PathVariable Long projectId){
        return projectService.sendApplication(Long.valueOf(principal.getName()), projectId)
                .switchIfEmpty(Mono.error(new NotFoundException("Not found!")));
    }

    @PostMapping("/invite/{projectId}")
    public Mono<InfoResponse> inviteInProject(@PathVariable Long projectId, @RequestBody AuthenticationResponse invitation){
        return projectService.addInProject(projectId, invitation.getId())
                .thenReturn(new InfoResponse(HttpStatus.OK,"Successful invitation"))
                .onErrorReturn(new InfoResponse(HttpStatus.BAD_REQUEST,"The invitation was not successful"));
    }

    ///////////////////////////////////////////
    //   ___    ____   __    ____ ______   ____
    //  / _ \  / __/  / /   / __//_  __/  / __/
    // / // / / _/   / /__ / _/   / /    / _/
    ///____/ /___/  /____//___/  /_/    /___/
    ///////////////////////////////////////////

    @DeleteMapping("/delete/{projectId}")
    public Mono<InfoResponse> deleteProject(@PathVariable Long projectId){
        return projectService.deleteProject(projectId)
                .thenReturn(new InfoResponse(HttpStatus.OK,"Success deleting"))
                .onErrorReturn(new InfoResponse(HttpStatus.BAD_REQUEST,"Delete is not success"));
    }

    @DeleteMapping("/delete/invite/{inviteId}")
    public Mono<InfoResponse> deleteInvite(@PathVariable Long inviteId){
        return projectService.deleteInvite(inviteId)
                .thenReturn(new InfoResponse(HttpStatus.OK,"Success deleting"))
                .onErrorReturn(new InfoResponse(HttpStatus.BAD_REQUEST,"Delete is not success"));
    }

    @DeleteMapping("/delete/request/{requestId}")
    public Mono<InfoResponse> deleteRequest(@PathVariable Long requestId){
        return projectService.deleteRequest(requestId)
                .thenReturn(new InfoResponse(HttpStatus.OK,"Success deleting"))
                .onErrorReturn(new InfoResponse(HttpStatus.BAD_REQUEST,"Delete is not success"));
    }

    @DeleteMapping("/kick/{projectId}")
    public Mono<InfoResponse> kickFromProject(@PathVariable Long projectId, @RequestBody AuthenticationResponse invitation){
        return projectService.kickFromProject(projectId, invitation.getId())
                .thenReturn(new InfoResponse(HttpStatus.OK,"Successful exclusion"))
                .onErrorReturn(new InfoResponse(HttpStatus.BAD_REQUEST,"The exception was not successful"));
    }

    ////////////////////////
    //   ___   __  __ ______
    //  / _ \ / / / //_  __/
    // / ___// /_/ /  / /
    ///_/    \____/  /_/
    ////////////////////////

    @PutMapping("/update/{projectId}")
    public Mono<InfoResponse> updateProject(@RequestBody ProjectDTO projectDTO, @PathVariable Long projectId){
        return projectService.updateProject(projectDTO, projectId)
                .thenReturn(new InfoResponse(HttpStatus.OK,"Success updating"))
                .onErrorReturn(new InfoResponse(HttpStatus.BAD_REQUEST,"Update is not success"));
    }

}

File: scrum-service/src/main/kotlin/com/tyiu/scrumservice/controller/ProjectController.kt
package com.tyiu.scrumservice.controller

import com.tyiu.ideas.model.dto.IdeaMarketDTO
import com.tyiu.scrumservice.model.*
import com.tyiu.scrumservice.service.ProjectService
import kotlinx.coroutines.reactor.asFlux
import org.springframework.http.HttpStatus
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.oauth2.jwt.Jwt
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono

@RestController
@RequestMapping("/api/v1/scrum-service/project")
class ProjectController (private val projectService: ProjectService) {

    @GetMapping("/all")
    @PreAuthorize("hasAnyRole('PROJECT_OFFICE' , 'ADMIN')")
    fun getAllProjects(): Flux<ProjectDTO> = projectService.getAllProjects().asFlux()

    @GetMapping("/private/all")
    @PreAuthorize("hasAnyRole('INITIATOR', 'MEMBER', 'TEAM_OWNER', 'ADMIN')")
    fun getYourProjects(@AuthenticationPrincipal jwt: Jwt): Flux<ProjectDTO> = projectService.getYourProjects(jwt.id).asFlux()

    @GetMapping("/active/all")
    @PreAuthorize("hasAnyRole('INITIATOR', 'MEMBER', 'TEAM_OWNER', 'ADMIN')")
    fun getYourActiveProjects(@AuthenticationPrincipal jwt: Jwt): Flux<ProjectDTO> = projectService.getYourActiveProjects(jwt.id).asFlux()

    @GetMapping("/{projectId}")
    //@PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'MEMBER', 'TEAM_OWNER', 'ADMIN')")
    suspend fun getOneProject(@PathVariable projectId: String): Mono<ProjectDTO> = projectService.getOneProject(projectId).toMono()

    @GetMapping("/members/{projectId}/all")
    @PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'MEMBER', 'TEAM_OWNER', 'ADMIN')")
    fun getProjectMembers(@PathVariable projectId: String): Flux<ProjectMemberDTO>? = projectService.getProjectMembers(projectId)?.asFlux()

    @GetMapping("/marks/{projectId}/all")
    @PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'MEMBER', 'TEAM_OWNER', 'ADMIN')")
    fun getProjectMarks(@PathVariable projectId: String): Flux<ProjectMarksDTO>? = projectService.getProjectMarks(projectId)?.asFlux()

    @PostMapping("/send")
    //@PreAuthorize("hasAnyRole('PROJECT_OFFICE' , 'ADMIN')")
    suspend fun createProject(@RequestBody ideaMarketDTO: IdeaMarketDTO): Flux<ProjectMember> = projectService.createProject(ideaMarketDTO).asFlux()

    @PostMapping("/{projectId}/add/members")
    //@PreAuthorize("hasAnyRole('TEAM_OWNER', 'ADMIN')")
    suspend fun addMembersInProject(@PathVariable projectId: String, @RequestBody teamMemberRequest: TeamMemberRequest): Mono<ProjectMember> = projectService.addMembersInProject(projectId,teamMemberRequest).toMono()

    @PostMapping("/{projectId}/add/marks")
    //@PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'ADMIN')")
    suspend fun addMarksInProject(@PathVariable projectId: String,@RequestBody projectMarksRequest: projectMarksRequest): Mono<ProjectMarks> = projectService.addMarksInProject(projectId, projectMarksRequest).toMono()

    @PutMapping("/{projectId}/status/change")
    //@PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'ADMIN')")
    suspend fun pauseProject(@PathVariable projectId: String) : Mono<InfoResponse> {
        return Mono.just(projectService.pauseProject(projectId))
            .thenReturn(InfoResponse(HttpStatus.OK,"Статус проекта изменён"))
            .onErrorReturn(InfoResponse(HttpStatus.BAD_REQUEST,"Статус проекта не был изменён"))
    }

    @PutMapping("/{projectId}/finish/change")
    //@PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'ADMIN')")
    suspend fun putFinishProject(@PathVariable projectId: String,@RequestBody projectFinishRequest: ProjectFinishRequest) : Mono<InfoResponse> {
        return Mono.just(projectService.putFinishProject(projectId,projectFinishRequest))
            .thenReturn(InfoResponse(HttpStatus.OK,"Проект успешно завершён"))
            .onErrorReturn(InfoResponse(HttpStatus.BAD_REQUEST,"Проект не был завершён"))
    }

    @PutMapping("/leader/change")
    //@PreAuthorize("hasAnyRole('INITIATOR', 'PROJECT_OFFICE', 'ADMIN')")
    suspend fun putTeamLeader(@RequestBody projectLeaderRequest: ProjectLeaderRequest) : Mono<InfoResponse> {
        return Mono.just(projectService.putTeamLeader(projectLeaderRequest))
            .thenReturn(InfoResponse(HttpStatus.OK,"Лидер команды назначён"))
            .onErrorReturn(InfoResponse(HttpStatus.BAD_REQUEST,"Лидер команды не был назначен"))
    }

}